text/0000-otlp-trace-data-format.md

```diff
@@ -0,0 +1,361 @@
+
+  // The `span_id` of this span's parent span. If this is a root span, then this
+  // field must be empty. The ID is an 8-byte array.
+  bytes parent_span_id = 4;
+
+  // An optional resource that is associated with this span. If not set, this span
+  // should be part of a ResourceSpan that does include the resource information, unless resource
+  // information is unknown.
+  Resource resource = 5;
+
+  // A description of the span's operation.
+  //
+  // For example, the name can be a qualified method name or a file name
+  // and a line number where the operation is called. A best practice is to use
+  // the same display name at the same call point in an application.
+  // This makes it easier to correlate spans in different traces.
+  //
+  // This field is semantically required to be set to non-empty string.
+  // When null or empty string received - receiver may use string "name"
+  // as a replacement. There might be smarted algorithms implemented by
+  // receiver to fix the empty span name.
+  //
+  // This field is required.
+  string name = 6;
+
+  // Type of span. Can be used to specify additional relationships between spans
+  // in addition to a parent/child relationship.
+  enum SpanKind {
+    // Unspecified. Do NOT use as default.
+    // Implementations MAY assume SpanKind to be INTERNAL when receiving UNSPECIFIED.
+    SPAN_KIND_UNSPECIFIED = 0;
+
+    // Indicates that the span is used internally. Default value.
+    INTERNAL = 1;
+
+    // Indicates that the span covers server-side handling of an RPC or other
+    // remote network request.
+    SERVER = 2;
+
+    // Indicates that the span covers the client-side wrapper around an RPC or
+    // other remote request.
+    CLIENT = 3;
+
+    // Indicates that the span describes producer sending a message to a broker.
+    // Unlike client and  server, there is no direct critical path latency relationship
+    // between producer and consumer spans.
+    PRODUCER = 4;
+
+    // Indicates that the span describes consumer receiving a message from a broker.
+    // Unlike client and  server, there is no direct critical path latency relationship
+    // between producer and consumer spans.
+    CONSUMER = 5;
+  }
+
+  // Distinguishes between spans generated in a particular context. For example,
+  // two spans with the same name may be distinguished using `CLIENT` (caller)
+  // and `SERVER` (callee) to identify queueing latency associated with the span.
+  SpanKind kind = 7;
+
+  // The start time of the span. On the client side, this is the time kept by
+  // the local machine where the span execution starts. On the server side, this
+  // is the time when the server's application handler starts running.
+  //
+  // This field is semantically required. When not set on receive -
+  // receiver should set it to the value of end_time field if it was
+  // set. Or to the current time if neither was set. It is important to
+  // keep end_time > start_time for consistency.
```

Comment:
if it's semantically required, then the message should be rejected if the field is missing.

Reply:
@yurishkuro I spent a bit time thinking about this and I think it should not be OTLP's business to make recommendations about how exactly to behave for certain invalid input data, like this one. The best behavior depends on the specific use case.
For example if this is received by a backend that needs to display the span visually then it may reject the data or attempt to insert a substitute "correct" value just for displaying purposes (and then indicate visually that it is "corrected").
If this is received by an intermediary node like Collector then the Collector most likely doesn't care about this particular field. The Collector can forward the Span as is and let the subsequent receivers decide what to do about it (and the Collector may record the fact that it received bad data and from whom, but will still forward it).
So, what I think is appropriate in this specification is to define what is the correct data but refrain from prescribing a behavior when incorrect data is received.